File: code/547.number-of-provinces.php
<?php
/*
 * @lc app=leetcode id=547 lang=php
 *
 * [547] Number of Provinces
 */

// @lc code=start
class Solution {
    /**
     * @param Integer[][] $isConnected
     * @return Integer
     */
    function findCircleNum($isConnected) {
        $province = 0;
        $visited = [];
        foreach ($isConnected as $i => $_) {
            if (!isset($visited[$i])) {
                $province++;
                $this->helper($i, $visited, $isConnected);
            }
        }

        return $province;
    }

    function helper($city, &$visited, $isConnected) {
        $visited[$city] = true;
        foreach ($isConnected[$city] as $i => $v) {
            if (isset($visited[$i])) continue;
            if ($v === 0) continue;
            $this->helper($i, $visited, $isConnected);
        }
    }
}
// @lc code=end

File: code/547.number-of-provinces.go
package code

/*
 * @lc app=leetcode id=547 lang=golang
 *
 * [547] Number of Provinces
 */

// @lc code=start
func findCircleNum(isConnected [][]int) int {
	result := 0
	visited := make([]bool, len(isConnected))

	for k, _ := range isConnected {
		if visited[k] {
			continue
		}

		result++
		rec547(k, &visited, isConnected)
	}

	return result
}

func rec547(k int, visited *[]bool, isConnected [][]int) {
	(*visited)[k] = true
	for column, connected := range isConnected[k] {
		if (*visited)[column] {
			continue
		}
		if connected == 0 {
			continue
		}
		rec547(column, visited, isConnected)
	}
}

// @lc code=end
